adjacentSquares.js

```diff
@@ -0,0 +1,105 @@
+/*
+series of squares, 2d plane
+side length of 1
+find size of largest group in adjacent squares
+each sqaure, x-y pair
+
+i: array of squares [[0, 0], [1, 1]]
+o: largest group of adajcent squares
+
+const squares = [
+    [0.5, 0.5],
+    [1.5, 0.5],
+    [2, 0.5],
+    [2.5, -0.5],
+    [2.5, -1.5],
+    [1.5, -1.5],
+];
+
+*/
+
+const SQUARE_SIDE_LENGTH = 1;
+
+const areConected = (nodeOne, nodeTwo) => {
+    const [nodeOneX, nodeOneY] = nodeOne;
+    const [nodeTwoX, nodeTwoY] = nodeTwo;
+
+    const differenceBetweenNodesX = Math.abs(nodeTwoX - nodeOneX);
+    const differenceBetweenNodesY = Math.abs(nodeTwoY - nodeOneY);
+
+    const areEqualedToSideLengthX =
+        differenceBetweenNodesX === 1 && differenceBetweenNodesY < SQUARE_SIDE_LENGTH;
+    const areEqualedToSideLengthY =
+        differenceBetweenNodesY === 1 && differenceBetweenNodesX < SQUARE_SIDE_LENGTH;
+
+    return areEqualedToSideLengthX || areEqualedToSideLengthY;
+};
+
+const buildGraph = (nodes) => {
+    const numNodes = nodes.length;
+    const graph = {};
+
+    for (let nodeOne = 0; nodeOne < numNodes; nodeOne++) {
+        for (let nodeTwo = nodeOne + 1; nodeTwo < numNodes; nodeTwo++) {
+            if (!areConected(nodes[nodeOne], nodes[nodeTwo])) continue;
+
+            const containsNodeOne = graph.hasOwnProperty(nodeOne);
+            const containsNodeTwo = graph.hasOwnProperty(nodeTwo);
+
+            if (!containsNodeOne) graph[nodeOne] = [];
+            if (!containsNodeTwo) graph[nodeTwo] = [];
+
+            graph[nodeOne].push(nodeTwo);
+            graph[nodeTwo].push(nodeOne);
+        }
+    }
+
+    return graph;
+};
+
+const getComponentSize = (startNode, graph, visited) => {
+    const queue = new Queue();
+
+    visited.add(startNode);
+    queue.enqueue(startNode);
+
+    let componentSize = 0;
+    while (queue.size() > 0) {
+        // Remove node
+        const node = queue.dequeue();
+
+        // Process node
+        componentSize++;
+
+        // Get neighbors
+        const containsNode = graph.hasOwnProperty(node);
+        if (!containsNode) continue;
+
+        const neighbors = graph[node];
+        for (const neighbor of neighbors) {
+            if (visited.has(neighbor)) continue;
+
+            visited.add(neighbor);
+            queue.enqueue(neighbor);
+        }
+    }
+
+    return componentSize;
+};
+
+const getLargestGroup = (squares) => {
+    const graph = buildGraph(squares);
+    const numSquares = squares.length;
+    const visitedSquare = new Set();
+    let maxGroupSquares = 0;
+
+    for (let square = 0; square < numSquares; square++) {
```

Comment:
P: Here, you're generating a unique identifier for each square. Why do we need one? And for each square, what is the unique identifier that you give it?

Reply:
For the unique identifier, we need it to distinguish each individual node from the array of nodes. We're also using the the index based numbers based off the input array. The graph is also built with the index identifier. That way we can access the node's neighbors.

Reply:
Yup, that's right! We could have two squares that contain exactly the same data, but they're actually different squares. So, indices are the only way for us to uniquely identify a specific square. That's why we use them in the `graph` object and in the `visited` set.